File: arr/leetcode_Merge_Intervals.java
/*Merge_Intervals

Given a collection of intervals, merge all overlapping intervals.
For example,
Given [1,3],[2,6],[8,10],[15,18],
return [1,6],[8,10],[15,18].
*/

import java.util.ArrayList;


class leetcode_Merge_Intervals {
    public static void main(String[] args) {
    }
    public static ArrayList<Interval> merge(ArrayList<Interval> intervals) {
        ArrayList<Interval> res = new ArrayList<Interval>();
        if (intervals.size() == 0) {
            return res;
        }
        res.add(intervals.get(0));
        for (int i = 1; i < intervals.size(); ++i) {
            insert(res, intervals.get(i));
        }
        return res;
    }
    public void insert(ArrayList<Interval> intervals, Interval newInterval) {
        int s = -1, e = -1;
        for (int i = 0; i < intervals.size(); i++) {
            if (s == -1 && intervals.get(i).end >= newInterval.start) {
                s = i;
            }
            if (intervals.get(i).start <= newInterval.end) {
                e = i;
            }
        }
        if (s == -1) {
            intervals.add(newInterval);
        } else if (e == -1) {
            intervals.add(0, newInterval);
        } else {
            int start = Math.min(intervals.get(s).start, newInterval.start);
            int end = Math.max(intervals.get(e).end, newInterval.end);
            intervals.subList(s, e + 1).clear();
            intervals.add(s, new Interval(start, end));
        }
    }
}


class Interval {
    int start;
    int end;
    Interval() {
        start = 0;
        end = 0;
    }
    Interval(int s, int e) {
        start = s;
        end = e;
    }
}


/* Python Version
def merge(self, intervals):
    ret = []
    for interval in intervals:
        ret = self.mergeOne(ret, interval)
    return ret

def mergeOne(self, intervals, newInterval):
    if not intervals:
        return [newInterval]
    
    if not newInterval:
        return intervals
    
    s = -1
    e = -1
    
    for i in xrange(len(intervals)):
        if s == -1 and intervals[i].end >= newInterval.start:
            s = i
        if intervals[i].start <= newInterval.end:
            e = i
    
    if s == -1:
        intervals.append(newInterval)
    elif e == -1:
        intervals.insert(0, newInterval)
    else:
        newStart = min(newInterval.start, intervals[s].start)
        newEnd = max(newInterval.end, intervals[e].end)
        
        del intervals[s : e + 1]
        intervals.insert(s, Interval(newStart, newEnd))
    return intervals
*/

File: arr/leetcode_Merge_Intervals.cpp
/**
 * Copyright 2016
 * author: madarfacar
 *
 * leetcode merge intervals
 */

#include <stdio.h>
#include <stdlib.h>
#include <vector>

using std::vector;


/**
 * Definition for an interval.
 */
struct Interval {
    int start;
    int end;
    Interval() : start(0), end(0) {}
    Interval(int s, int e) : start(s), end(e) {}
};


class Solution {
 public:
  vector<Interval> merge(const vector<Interval>& intervals) {
    vector<Interval> merged;
    for (Interval interval : intervals) {
      insert(&merged, interval);
    }
    return merged;
  }

 private:
  void insert(vector<Interval>* intervals, Interval newInterval) {
    int inserted_begin = -1;
    int inserted_end = -1;

    for (int i = 0; i < intervals->size(); ++i) {
      if (inserted_begin == -1 && newInterval.start <= (*intervals)[i].end) {
        inserted_begin = i;
      }

      if (newInterval.end >= (*intervals)[i].start) {
        inserted_end = i;
      }
    }

    if (inserted_begin == -1) {
      intervals->push_back(newInterval);
    } else if (inserted_end == -1) {
      intervals->insert(intervals->begin(), newInterval);
    } else {
      int merged_start = std::min(newInterval.start, (*intervals)[inserted_begin].start);
      int merged_end = std::max(newInterval.end, (*intervals)[inserted_end].end);
      Interval merged (merged_start, merged_end);

      intervals->erase(intervals->begin() + inserted_begin, intervals->begin() + inserted_end + 1);
      intervals->insert(intervals->begin() + inserted_begin, merged);
    }
  }
};


int main() {
  Solution sol;
  vector<Interval> intervals {
    Interval (1, 3),
    Interval (2, 6),
    Interval (8, 10),
    Interval (15, 18),
  };

  vector<Interval> inserted = sol.merge(intervals);

  for (Interval interval : inserted) {
    printf("start: %d end: %d\n", interval.start, interval.end);
  }
}
